core/src/main/java/jenkins/slaves/restarter/UnixSlaveRestarter.java

```diff
@@ -57,11 +59,61 @@ public void restart() throws Exception {
         }
 
         // exec to self
-        String exe = Daemon.getCurrentExecutable();
+        String exe = getCurrentExecutable();
         LIBC.execv(exe, new StringArray(args.toArray(new String[0])));
         throw new IOException("Failed to exec '" + exe + "' " + LIBC.strerror(Native.getLastError()));
     }
 
+    /**
+     * Gets the current executable name.
+     */
+    private static String getCurrentExecutable() {
+        int pid = LIBC.getpid();
+        String name = "/proc/" + pid + "/exe";
+        File exe = new File(name);
+        if (exe.exists()) {
+            try {
+                String path = resolveSymlink(exe);
+                if (path != null) {
+                    return path;
+                }
+            } catch (IOException e) {
+                LOGGER.log(FINE, "Failed to resolve symlink " + exe, e);
+            }
+            return name;
+        }
+
+        // cross-platform fallback
+        return System.getProperty("java.home") + "/bin/java";
+    }
+
+    private static String resolveSymlink(File link) throws IOException {
+        String filename = link.getAbsolutePath();
+
+        for (int sz = 512; sz < 65536; sz *= 2) {
+            Memory m = new Memory(sz);
+            int r = LIBC.readlink(filename, m, new NativeLong(sz));
+            if (r < 0) {
+                int err = Native.getLastError();
+                if (err == 22 /*EINVAL --- but is this really portable?*/) {
+                    return null; // this means it's not a symlink
+                }
+                throw new IOException(
+                        "Failed to readlink " + link + " error=" + err + " " + LIBC.strerror(err));
+            }
+
+            if (r == sz) {
+                continue; // buffer too small
+            }
+
+            byte[] buf = new byte[r];
+            m.read(0, buf, 0, r);
+            return new String(buf);
+        }
+
+        throw new IOException("Failed to readlink " + link);
+    }
+
```

Comment:
Compare with [the original](https://github.com/kohsuke/akuma/blob/9d832605eef753e895c8125e7d32bd4917e4d19a/src/main/java/com/sun/akuma/Daemon.java#L220-L264).